### PR TITLE
Add test proving that audit logs are immune to sampling

### PR DESCRIPTION
**Pull request by vrto**
## Description

Telemetry sampling is a native observability concept that helps the emitter to reduce the volume of telemetry data to a manageable level. This is usually fine as long as we're not talking audit log - we don't want to drop any actions that will appear in the security audit log.

## Test plan

Automated tests.

### CODE REVIEW

File: internal/audit/audit_test.go [hunk 1]
```diff
@@ -144,3 +145,23 @@ func TestLog(t *testing.T) {
 		})
 	}
 }
+
+func TestSampling(t *testing.T) {
```
Comment (vrto) [inline internal/audit/audit_test.go:149]: @jhchabran @bobheadxi do you have any ideas how to make this test fail?

`SRC_LOG_SAMPLING_INITIAL` env var [should control the sampling](https://docs.sourcegraph.com/admin/observability/logs#log-sampling), but it is only read during the [initialization phase](https://github.com/sourcegraph/log/blob/main/sinks_output.go#L46).

Setting it in my test made no difference, as the log library isn't initialized.

My intention is to make the test fail, and then I _assume_ the fix may be adding  a unique ID to each audit log entry, so the entries aren't ["identical"](https://docs.sourcegraph.com/admin/observability/logs#log-sampling):

> The first 100 identical log entries per second will always be output, but thereafter only every 100th identical message will be output.

Comment (jhchabran) [reply internal/audit/audit_test.go:149]: You can't really do this without re-running the test yourself. You could simply shell out and run `SRC_LOG_SAMPLINGI=... go test -run Test...` and check the exit code.

Comment (jhchabran) [reply internal/audit/audit_test.go:149]: You can make it automated: 

```go
var unwrapped = flag.Bool("unwrapped", false, "bla bla")

func TestWrapped() { 
  os.Exec("go", "run", ..., "-run", "TestUnwrapped", "-unwrapped")
  // ...
}

func TestUnwrapped() { 
  if !*unwrapped { 
    t.Skip("not running this unless asked to")
  }
  // your thing 
}
```

Comment (jhchabran) [reply internal/audit/audit_test.go:149]: Some example https://jamiethompson.me/posts/Unit-Testing-Exec-Command-In-Golang/

Comment (vrto) [reply internal/audit/audit_test.go:149]: @bobheadxi @burmudar  setting the env var before running the tests still won't make them fail (I'm trying the approach @jhchabran suggest above - `SRC_LOG_SAMPLING_INITIAL=5 go test -run TestSampling`), but it's still all good and happy, no logs get sampled. Some init code must be missing. Do you know if calling [logtest.Captured(t)](https://docs.sourcegraph.com/dev/how-to/add_logging#testing-usage) is enough for initializing the sampling logic in `sourcegraph/log` library?

Comment (vrto) [reply internal/audit/audit_test.go:149]: Note to self: sampling works when running the app via `sg start`, trying to figure out which part of the logging setup gets called during `sg start`, but not in the tests

Comment (vrto) [reply internal/audit/audit_test.go:149]: Note to self: sampling isn't configured in neither `logtest.Init` nor `logtest.Captured`, only in `sglog.Init`. It's oddly coupled to the [build sinks logic](https://sourcegraph.com/-/editor?remote_url=git%40github.com%3Asourcegraph%2Flog.git&branch=main&file=sinks_output.go&start_row=44&start_col=24&end_row=44&end_col=24&editor=JetBrains&version=v2.0.1) - I'm not sure if that was intentional or not at this time.

**Comment by vrto** [comment]
Closing, addressed in https://github.com/sourcegraph/sourcegraph/pull/42427

**Pull request closed without merging**